### PR TITLE
Send user updated websocket event on plugin.UpdateUser

### PR DESCRIPTION
**Pull request by chetanyakan**
#### Summary

- Sends a WebSocket event when `plugin.UpdateUser` is called so it is updated immediately for all the users.

#### Release Note

```release-note
- UserUpdate WebSocket Event is now broadcast by 2 more APIs - plugin.UpdateUser and ConvertBotToUser
- Change in a User's NotifyProps during the user import process will not send an email notification. This is done to make it consistent with other parts of the import process where change in user's attributes would also not send any notifications.
```

### CODE REVIEW

File: app/plugin_api.go
Comment (agnivade) [inline app/plugin_api.go:287]: We should move this inside the `UpdateUser` method itself. There are 2 places where `sendUpdatedUserEvent` is called after UpdateUser. We should remove those as well.

**Review by agnivade — CHANGES_REQUESTED** [review]
Thanks @chetanyakan. Please take a look at the comments.

File: app/plugin_api.go
Comment (chetanyakan) [reply app/plugin_api.go:287]: @agnivade This will send the UpdatedUser WebSocket event in other places a well like - 
`[UpdateUserNotifyProps](https://github.com/mattermost/mattermost-server/blob/c1dd23a3c80365c4ca4e2721e55c0c2c90038b25/app/user.go#L1348)` and `[ConvertBotToUser](https://github.com/mattermost/mattermost-server/blob/c1dd23a3c80365c4ca4e2721e55c0c2c90038b25/app/user.go#L2334)`

Would that be an acceptable change?

Comment (agnivade) [reply app/plugin_api.go:287]: Yes, I think that should be fine. Thanks.

Comment (chetanyakan) [reply app/plugin_api.go]: @agnivade done

**Comment by agnivade** [comment]
Woops, @chetanyakan - please update the PR with the latest master. Thanks!

**Comment by agnivade** [comment]
Also, let's flesh out the release notes properly by specifically mentioning which 2 API requests will now broadcast the user update messages via websocket.

**Comment by chetanyakan** [comment]
Done @agnivade

**Review by agnivade — APPROVED** [review]

**Review by agarciamontoro — APPROVED** [review]
Thank you for your contribution, @chetanyakan :tada:

(and sorry for the late review, I was out on vacation!)

**Comment by lindy65** [comment]
Hi @agnivade / @agarciamontoro - is there a Jira ticket for this PR? Please link me to it? Thanks!

**Comment by agarciamontoro** [comment]
Hm, that's a good question I don't an answer for, I had not noticed when I reviewed it. Maybe @agnivade or @chetanyakan know.

**Comment by agnivade** [comment]
There is no JIRA ticket @lindy65. This is something @chetanyakan created on his own. I am not sure of the QA process we follow for contributor generated PRs.

**Comment by lindy65** [comment]
Hi @agnivade - would you help create a Jira ticket please?  This is required so that once the PR is merged, we don't lose track of it and have something to test against from resolved tickets.  Please also help add test steps to the ticket? TIA :)

**Review by agnivade — CHANGES_REQUESTED** [review]
Actually, @chetanyakan - apologies for asking you to do this again, but on a closer look, I realized that we don't want to do this for UpdateNotifyProps. That's because it happens during the user import process and there is no need to send any notifications.

It should not be too much work. The method is just used in 2 places. We need to add another flag to `UpdateUserNotifyProps` for `sendNotifications` and pass false from app/import_functions.go.

And only send the user event if `sendNotifications` is set to true.

Everything else remains same. Sorry for the double work here.

**Comment by agnivade** [comment]
@lindy65 - https://mattermost.atlassian.net/browse/MM-31915 is the ticket, but please hold off from testing for now. The PR needs to be updated a bit.

**Comment by lindy65** [comment]
Thanks @agnivade - I'll hold off testing until you give me the go-ahead.  Can you tell me what plugin updates a user profile so I can test using it?

**Comment by agnivade** [comment]
There are many like that but probably @agarciamontoro knows a suitable one to test this.

**Comment by agarciamontoro** [comment]
I cannot think of one right now, actually :thinking: I checked that the Github plugin does call UpdateUser when disconnecting the Github account: https://github.com/mattermost/mattermost-plugin-github/blob/8f2addacdcf24636d697fc2dc015cbadbcbb2c8a/server/plugin/plugin.go#L331, but I'm not sure if that's very userful for your testing purposes, @lindy65.

**Comment by lindy65** [comment]
Thanks @agarciamontoro :)  Perhaps I should test this after merge where it'll be easier with available plugins on `master`

**Comment by agnivade** [comment]
Hey @chetanyakan - how is it going with this one? Let us know if you need any help.

**Comment by chetanyakan** [comment]
Sorry, @agnivade, I have been really busy. Will update this shortly.

**Comment by chetanyakan** [comment]
> Actually, @chetanyakan - apologies for asking you to do this again, but on a closer look, I realized that we don't want to do this for UpdateNotifyProps. That's because it happens during the user import process and there is no need to send any notifications.
> 
> It should not be too much work. The method is just used in 2 places. We need to add another flag to `UpdateUserNotifyProps` for `sendNotifications` and pass false from app/import_functions.go.
> 
> And only send the user event if `sendNotifications` is set to true.
> 
> Everything else remains same. Sorry for the double work here.

Hey @agnivade just to clarify, I will be making the following changes:

1. Update the `app.UpdateUser` method to call `a.sendUpdatedUserEvent(*user)` only when the value of the second parameter `sendNotifications` is true.
2. Update the method `app.UpdateUserNotifyProps` to add an extra parameter `sendNotifications` and pass that to `UpdateUser`.
3. Update the usages of `UpdateUserNotifyProps` method to pass the values of  `sendNotifications` and pass `false` in `app/import_functions.go`.

Is that correct?

**Comment by agnivade** [comment]
That is correct! A small note there is that `app/import_functions.go` is the only place that calls `UpdateUserNotifyProps`.

I should add here for posterity that this change will make `UpdateUserNotifyProps` not send mail notifications from import_functions.go. But there are other areas, where `UpdateUser` / `UpdateUserRoles` are called during the import process with notifications disabled. And it sounds like the correct process to me. So we would be actually making the behavior more consistent.

And for full transparency, we should add this in release notes too, that change in a user's notify props during the import process will not send a mail notification. This is done to make it consistent with changes in other attributes of the user.

**Comment by agnivade** [comment]
Are the release notes still correct? `UpdateUserNotifyProps` won't broadcast an event any more correct?

/update-branch

**Comment by agnivade** [comment]
Woops, please update the PR with the latest master.

**Comment by chetanyakan** [comment]
@agnivade I have updated the PR to the head of `master` and updated the release notes.

Your note was correct, `UpdateUserNotifyProps` won't broadcast the user updated event.

**Review by agnivade — APPROVED** [review]
Nice work @chetanyakan !

**Comment by agnivade** [comment]
@lindy65 - over to you :)

**Comment by lindy65** [comment]
> I cannot think of one right now, actually 🤔 I checked that the Github plugin does call UpdateUser when disconnecting the Github account: https://github.com/mattermost/mattermost-plugin-github/blob/8f2addacdcf24636d697fc2dc015cbadbcbb2c8a/server/plugin/plugin.go#L331, but I'm not sure if that's very userful for your testing purposes, @lindy65.

Hey @agnivade - would this PR be better tested after merge based on what `@agarciamontoro` mentioned above?  I'm not sure which plugin tests this fix properly but there is more choice on `community-daily`? Thoughts?

**Comment by agnivade** [comment]
I agree.

**Review by lindy65 — APPROVED** [review]
This will be tested on `community-daily` after merge 👍

**Pull request merged**
